src/compiler/visitorPublic.ts

```diff
@@ -159,29 +212,29 @@ namespace ts {
 
         // Visit each original node.
         for (let i = 0; i < count; i++) {
-            const node: T = nodes[i + start];
-            const visited = node !== undefined ? visitor(node) : undefined;
-            if (updated !== undefined || visited === undefined || visited !== node) {
+            const node = nodes[i + start];
+            const visited = node !== undefined ? (visitor ? visitor(node) : node) : undefined;
+            if (updated !== undefined || visited === undefined || visited !== node as VisitResult<Node>) {
                 if (updated === undefined) {
                     // Ensure we have a copy of `nodes`, up to the current index.
                     updated = nodes.slice(0, i);
                 }
                 if (visited) {
                     if (isArray(visited)) {
                         for (const visitedNode of visited) {
-                            void Debug.assertNode(visitedNode, test);
-                            updated.push(visitedNode as T);
+                            updated.push(visitedNode);
                         }

```

Comment:
If we actually don't want to test each, this could be `updated.push(...visited)`.